server/chat/backend/agent/prompt/provider_rules.py

```
@@ -59,10 +59,23 @@ def build_provider_constraints(provider_preference: Optional[Any]) -> Tuple[str,
     return provider_text, provider_restrictions, provider_constraints
 
 
+def get_gcp_disabled_projects(user_id: Optional[str]) -> List[str]:
+    """Return SA-mode disabled-project IDs for ``user_id`` (empty if none/unknown)."""
+    if not user_id:
+        return []
+    try:
+        from utils.auth.stateless_auth import get_user_preference
+        disabled = get_user_preference(user_id, "gcp_sa_disabled_projects", default=[]) or []
+        return [pid for pid in disabled if isinstance(pid, str) and pid]
+    except Exception:
```

Ruff (0.15.14)

[warning] Do not catch blind exception: `Exception`

(BLE001)